### PR TITLE
Ability to bypass usage of `litellm.Router`

### PR DESCRIPTION
**Pull request by jamesbraza**
There are multiple outstanding issues with LiteLLM's `Router` as of `litellm==1.48.10`:

1. It doesn't support TPM: https://github.com/BerriAI/litellm/issues/5783
2. The `Router` has several race conditions that keep killing our training runs

https://github.com/Future-House/paper-qa/pull/520 in-housed rate limits management, which was most of the reason we adopted the `Router` in https://github.com/Future-House/paper-qa/pull/315. Now the `Router` just enables model grouping, which we don't really use internally at FutureHouse. Paired with the `Router`'s race conditions, we decided to make usage of the `Router` optional.

So, this PR:
- Creates a back door to bypass `Router` usage, instead using LiteLLM's free functions
- Removes "router" mentions from the code base where stuff is actually just "model list" specific

### CODE REVIEW

File: paperqa/llms.py [hunk 1]
```diff
@@ -510,21 +518,37 @@ def get_litellm_retrying_config(timeout: float = 60.0) -> dict[str, Any]:
...
+    config: dict = Field(
+        default_factory=dict,
+        description=(
+            "Configuration of this model containing several important keys. The"
+            " optional `model_list` key stores a list of all model configurations"
+            " (SEE: https://docs.litellm.ai/docs/routing). The optional"
+            " `router_kwargs` key is keyword arguments to pass to the Router class."
+            " Inclusion of a key `pass_through_router` with a truthy value will lead"
+            " to using not using LiteLLM's Router, instead just LiteLLM's free"
+            f" functions (see {PassThroughRouter.__name__}). The optional `rate_limit`"
```
Comment (mskarlin) [inline paperqa/llms.py:546]: We might note that the rate limiter will only work if using the router based config.

File: paperqa/llms.py [hunk 2]
```diff
@@ -622,19 +652,22 @@ async def acomplete_iter(  # type: ignore[override]
...
         self, messages: Iterable[dict[str, str]]
     ) -> Chunk:
-        response = await self.router.acompletion(self.name, messages)
+        response = await self.router.acompletion(self.name, list(messages))
         return Chunk(
-            text=response.choices[0].message.content,
-            prompt_tokens=response.usage.prompt_tokens,
-            completion_tokens=response.usage.completion_tokens,
+            text=cast(litellm.Choices, response.choices[0]).message.content,
+            prompt_tokens=response.usage.prompt_tokens,  # type: ignore[attr-defined]
```
Comment (mskarlin) [inline paperqa/llms.py:658]: these are defined even without a router, right?

File: paperqa/llms.py [hunk 3]
```diff
@@ -622,19 +652,22 @@ async def acomplete_iter(  # type: ignore[override]
...
         )
 
     @rate_limited
     async def achat_iter(  # type: ignore[override]
         self, messages: Iterable[dict[str, str]]
     ) -> AsyncIterable[Chunk]:
         completion = await self.router.acompletion(
-            self.name, messages, stream=True, stream_options={"include_usage": True}
+            self.name,
+            list(messages),
```
Comment (mskarlin) [inline paperqa/llms.py:668]: I think if messages is a generator here, this will pull every value into the list which might not be desired here.

**Review by mskarlin — APPROVED** [review]
Couple minor comments, looks great!

File: paperqa/llms.py (hunk 3 above)
Comment (jamesbraza) [reply paperqa/llms.py:668]: The issue is `litellm` types with `list[Message]`, so I was adhering to their type.

Earlier, we didn't type `self.router`, so this snuck under the radar

Comment (jamesbraza) [reply paperqa/llms.py:668]: What do you think we should do here?

File: paperqa/llms.py (hunk 1 above)
Comment (jamesbraza) [reply paperqa/llms.py]: Nice suggestion, added a note here about this. We discussed in person, but rate limits apply to both cases

File: paperqa/llms.py (hunk 3 above)
Comment (jamesbraza) [reply paperqa/llms.py:669]: We discussed this in-person

File: paperqa/llms.py (hunk 2 above)
Comment (jamesbraza) [reply paperqa/llms.py:659]: Yes they are, and I have two different unit tests now proving it 😈

**Comment by krrishdholakia** [comment]
> The Router has several race conditions that keep killing our training runs

Hi @jamesbraza what are these race conditions?

**Pull request merged**